IdentityCore/src/logger/MSIDLogger.m

```diff
@@ -156,74 +156,88 @@ - (void)logWithLevel:(MSIDLogLevel)level
+        }
+        
+        __auto_type threadName = [[NSThread currentThread] isMainThread] ? @" (main thread)" : nil;
+        if (!threadName) {
+            threadName = [NSThread currentThread].name ?: @"";
+        }
+        
+        __auto_type threadInfo = [[NSString alloc] initWithFormat:@"TID=%llu%@", tid, threadName];
+        
+        if (self.nsLoggingEnabled)
+        {
+            NSString *logLevelStr = [self stringForLogLevel:self.level];
             
-            __auto_type threadName = [[NSThread currentThread] isMainThread] ? @" (main thread)" : nil;
-            if (!threadName) {
-                threadName = [NSThread currentThread].name ?: @"";
-            }
+            NSString *log = [NSString stringWithFormat:@"%@ %@ %@ %@ [%@%@]%@ %@:%@ %@", threadInfo, sdkName, sdkVersion, [MSIDDeviceId deviceOSId], dateStr, correlationIdStr, componentStr, logLevelStr, sourceInfo, message];
             
-            __auto_type threadInfo = [[NSString alloc] initWithFormat:@"TID=%llu%@", tid, threadName];
+            NSLog(@"%@", log);
+        }
+        
+        if (self.callback || self.loggerConnector)
+        {
+            NSString *log = [NSString stringWithFormat:@"%@ %@ %@ %@ [%@%@]%@%@ %@", threadInfo, sdkName, sdkVersion, [MSIDDeviceId deviceOSId], dateStr, correlationIdStr, componentStr, sourceInfo, message];
             
-            if (self.nsLoggingEnabled)
-            {
-                NSString *logLevelStr = [self stringForLogLevel:self.level];
-                
-                NSString *log = [NSString stringWithFormat:@"%@ %@ %@ %@ [%@%@]%@ %@:%@ %@", threadInfo, sdkName, sdkVersion, [MSIDDeviceId deviceOSId], dateStr, correlationIdStr, componentStr, logLevelStr, sourceInfo, message];
-                
-                NSLog(@"%@", log);
-            }
+            BOOL lineContainsPII = self.piiLoggingEnabled ? containsPII : NO;
             
-            if (self.callback || self.loggerConnector)
+            if (self.callback) self.callback(level, log, lineContainsPII);
+            if (self.loggerConnector) [self.loggerConnector onLogWithLevel:level lineNumber:lineNumber function:function message:log];
+        }
+    };
+    
+    BOOL loggingQueueEnabled = YES;
+    if (self.loggerConnector) loggingQueueEnabled = self.loggerConnector.loggingQueueEnabled;
+    
+    if (loggingQueueEnabled)
+    {
+        // Prevent queue from growing infinitely large.
+        dispatch_semaphore_wait(self.queueSemaphore, DISPATCH_TIME_FOREVER);
+        
+        dispatch_async(self.loggerQueue, ^{
+            @autoreleasepool
             {
-                NSString *log = [NSString stringWithFormat:@"%@ %@ %@ %@ [%@%@]%@%@ %@", threadInfo, sdkName, sdkVersion, [MSIDDeviceId deviceOSId], dateStr, correlationIdStr, componentStr, sourceInfo, message];
+                logBlock();
```

Comment:
I think it should be fine, but just take a moment to double check the memory management impact since now you have an additional block you're capturing.

Reply:
verified, no memory impact.